qbee-agent-upgrade/qbee-agent-upgrade.sh

```diff
@@ -0,0 +1,242 @@
+  local pkg_arch
+  system_arch="$1"
+  pkg_arch="$2"
+
+  if [[ "$system_arch" != "$pkg_arch" ]]; then
+    echo "ERROR: system arch $system_arch is not the same as package arch $pkg_arch"
+    exit 1
+  fi
+}
+
+check_package_version() {
+  local existing_version
+  local new_version
+  existing_version="$1"
+  new_version="$2"
+
+  if [[ ! "$existing_version" =~ ^(2023\.[0-9]+|1\.2\.[0-9]+)$ ]]; then
+    echo "ERROR: upgrade from version $existing_version is not supported with this script"
+    exit 1
+  fi
+
+  if [[ "$new_version" =~ ^1\.[0-9]\.[0-9]+$ ]]; then
+    echo "ERROR: downgrade to version $new_version is not supported with this script"
+    exit 1
+  fi
+
+  if [[ "$existing_version" == "$new_version" ]]; then
+    echo "ERROR: existing version $existing_version is the same as new version $new_version"
+    exit 1
+  fi
+}
+
+check_dpkg_lock() {
+  if ls -lt /proc/[0-9]*/fd 2> /dev/null | grep -q /var/lib/dpkg/lock; then
+    echo "ERROR: Package manager is already running, unable to upgrade" 
+    exit 1
+  fi
+}
+
+check_yum_lock() {
+  if [[ -f /var/run/yum.pid ]]; then
+    echo "ERROR: Package manager is already running, unable to upgrade"
+    exit 1
+  fi
+}
+
+sanity_deb() {
+  local pkg
+  pkg="$1"
+
+  check_dpkg_lock
+
+  pkg_name=$(dpkg -I "$pkg" | awk '/^[ ]*Package:/{print $NF}')
+
+  check_package_name "$pkg_name"
+
+  system_arch=$(dpkg --print-architecture)
+  pkg_arch=$(dpkg -I "$pkg" | awk '/^[ ]*Architecture:/{print $NF}')
+
+  check_package_arch "$system_arch" "$pkg_arch"
+
+  existing_version=$(dpkg -s qbee-agent | awk '/^[ ]*Version:/{print $NF}')
+  new_version=$(dpkg -I "$pkg" | awk '/^[ ]*Version:/{print $NF}')
+
+  check_package_version "$existing_version" "$new_version"
+}
+
+sanity_rpm() {
+  local pkg
+  pkg="$1"
+
+  check_yum_lock
+
+  pkg_name=$(rpm --queryformat='%{name}\n' "$pkg")
+
+  check_package_name "$pkg_name"
+
+  pkg_arch=$(rpm --queryformat='%{arch}\n' -qp "$pkg")
+  system_arch=$(rpm --eval '%{_arch}')
+
+  check_package_arch "$system_arch" "$pkg_arch"
+
+  existing_version=$(rpm --queryformat='%{version}\n' -q qbee-agent)
+  new_version=$(rpm --queryformat='%{version}\n' -qp "$pkg")
+
+  check_package_version "$existing_version" "$new_version"
+  
+}
+
+# resolve package manager
+dpkg_path=$(command -v dpkg || true)
+rpm_path=$(command -v rpm || true)
+
+check_system_utilities
+
+if [[ -n "${dpkg_path}" ]]; then
+  sanity_deb "$1"
+elif [[ -n "${rpm_path}" ]]; then
+  sanitiy_rpm "$1"
+else
+  echo "ERROR: package manager not supported"
+  exit 1
+fi
+
+# Passed the sanity checking, we can now upgrade the package
+UPGRADE_SCRIPT=$(mktemp /tmp/qbee-agent-do-upgrade.XXXXXXXXX.sh)
+cat <<EOF > "${UPGRADE_SCRIPT}"
+#!/usr/bin/env bash
+set -ue
+EOF
+
+if [[ -n "${dpkg_path}" ]]; then
+
+  cat <<'EOF' >> "${UPGRADE_SCRIPT}"
+# Wait for dpkg lock to be released, there could be apt/dpkg processes running
+while ls -lt /proc/[0-9]*/fd 2> /dev/null | grep -q /var/lib/dpkg/lock; do
```

Comment:
Should we put some hard limit on number of iterations (e.g. max 30 minutes or 1800 iterations) and return an error if the lock isn't released on time?

(same for Yum lock check below)

Reply:
This script runs detached in the background, so any error returned will not be caught by qbee-agent. The lock is also checked as part of the sanity check, so this is just an extra safety measure in case there's any system package operations started after sanity check, but before the detached script snippet is executed (like unattended-upgrades etc.)
I think trying indefinitely would be ok here, but I can change if you think this is very bad practice.